core/imds/src/main/java/software/amazon/awssdk/imds/internal/DefaultEc2Metadata.java

```diff
@@ -152,41 +154,57 @@ public String toString() {
-                                        .build();
+        for (int tries = 1 ; tries <= retryPolicy.numRetries() ; tries ++) {
+
+            try {
+                String token = getToken();
+                URI uri = URI.create(endpoint + path);
+                HttpExecuteRequest httpExecuteRequest = REQUEST_MARSHALLER.createDataRequest(uri, SdkHttpMethod.GET, token,
+                                                                                             tokenTtl);
+                HttpExecuteResponse response = httpClient.prepareRequest(httpExecuteRequest).call();
+                int statusCode = response.httpResponse().statusCode();
+                Optional<AbortableInputStream> responseBody = response.responseBody();
+
+                if (statusCode == HttpURLConnection.HTTP_OK && responseBody.isPresent()) {
+                    abortableInputStream = responseBody.get();
+                    String data = IoUtils.toUtf8String(abortableInputStream);
+                    metadataResponse = new MetadataResponse(data);
+                } else if (statusCode == HttpURLConnection.HTTP_NOT_FOUND) {
+                    throw SdkServiceException.builder()
+                                             .message("The requested metadata at path ( " + path + " ) is not found ").build();
+                } else if (statusCode == HttpURLConnection.HTTP_OK) {
+                    throw SdkClientException.builder()
+                                            .message("Response body empty with Status Code " + statusCode).build();
+                } else {
+                    throw SdkClientException.builder()
+                                            .message("Instance metadata service returned unexpected status code " + statusCode)
+                                            .build();
+                }
+            } catch (SdkServiceException sd) {
+                throw SdkServiceException.builder().message(sd.getMessage()).cause(sd).build();
+            } catch (IOException | SdkClientException io) {
+
+                log.warn("Received an IOException {0} ", io);
+                if (tries == 3) {
+                    throw SdkClientException.builder().message("Unable to contact EC2 metadata service.").cause(io).build();
+                }
+
+                Duration backoffTime = retryPolicy.backoffStrategy()
+                                       .computeDelayBeforeNextRetry(RetryPolicyContext.builder()
+                                                                                      .retriesAttempted(tries - 1)
+                                                                                      .build());
+
+                try {
+                    Thread.sleep(backoffTime.toMillis());
+                } catch (InterruptedException e) {
+                    Thread.currentThread().interrupt();
+                }
+            } finally {
+                IoUtils.closeQuietly(abortableInputStream, log);
             }
```

Comment:
Can you break up this block(line 159 - 206 in case multi-line comment is not working) a little bit? It looks like it's doing retry, sending request and handling response. Ideally, a method should just do one thing.
Per Clean Code,
>The first rule of functions is that they should be small.

Reply:
Noted. Will modularize the code.